core/src/main/java/io/ddf/Factor.java

```diff
@@ -220,4 +239,83 @@ public boolean isOrdered() {
   public void setOrdered(boolean isOrdered) {
     this.mIsOrdered = isOrdered;
   }
+
+  public static class FactorBuilder {
```

Comment:
Why is this builder class is not the only way to initiate the Factor class? It is also not including some fields like `isOrdered`. Looks like we should remove the other public constructors of Factor and make this the only way to create a new Factor? Also we can make Factor an immutable class by moving `computeLevelMap` and `setLevels` from Factor to this builder. That will make Factor truly thread safe.